Modify context related data structures for Unix AMD64

janvorli opened the pull request:
This change modifies PAL_LIMITED_CONTEXT, StackFrameIterator and also all related code for Unix AMD64 builds so that they correspond to the Unix AMD64 calling convention.

janvorli commented:
@jkotas can you take a look please? This is the first one in a series of changes that will enable exception handling on Unix AMD64.

jkotas commented:
LGTM @smosier Could you please take a look as well?

smosier commented on src/Native/Runtime/EHHelpers.cpp:
```diff
@@ -133,8 +133,12 @@ COOP_PINVOKE_HELPER(void, RhpCopyContextFromExInfo,
     pContext->Rip = pPalContext->IP;
     pContext->Rsp = pPalContext->Rsp;
     pContext->Rbp = pPalContext->Rbp;
+#ifndef UNIX_AMD64_ABI
```
I'm not sure I'm a fan of #ifdefing under the _AMD64_ block -- I think I'd rather see us treat amd64 unix as aseparate 'target cpu' and have the whole block separated per platform, i.e. one for _AMD64_ && UNIX_AMD64_ABI  and one for _AMD64_ && !UNIX_AMD64_ABI

smosier commented on src/Native/Runtime/EHHelpers.cpp:
```diff
@@ -133,8 +133,12 @@ COOP_PINVOKE_HELPER(void, RhpCopyContextFromExInfo,
     pContext->Rdi = pPalContext->Rdi;
     pContext->Rsi = pPalContext->Rsi;
+#else // !UNIX_AMD64_ABI
+    pContext->Rdx = pPalContext->Rdx;
```
I might also encourage a quick refresher comment for the reader on the amd64 unix abi.  I.e. for windows, it would be something like 
`// param regs: rcx, rdx, r8, r9, scratch: rax (return val), preserved: rbp, rdi, rsi, rbx, r12-r15 `

smosier commented on src/Native/Runtime/amd64/AsmOffsetsCpu.h:
```diff
@@ -8,32 +8,49 @@
 //
 // NOTE: the offsets MUST be in hex notation WITHOUT the 0x prefix
 
+#ifdef UNIX_AMD64_ABI
```
with the amount of #ifdefing in here, does it make any sense to create another variant of the file here, whether it be a full amd64_unix subdirectory sibling of 'runtime\amd64' or as a windows/unix split within 'runtime\amd64'?

janvorli commented:
@smosier I've addressed your feedback. Can you please take another look?
Regardsing the comment describing the register usage, I've put it to the PAL_LIMITED_CONTEXT.
As for the AsmOffsetsCpu.h, I have made UNIX_AMD64_ABI stuff completely separate instead of the interleaving, but kept it in the same file for now.

jkotas commented:
@dotnet-bot test Ubuntu Debug please

jkotas commented:
@dotnet-bot test this please

jkotas commented:
@dotnet-bot test this please

janvorli commented:
Implements part of #911
